### PR TITLE
Fix for CoreCLR issue 2881 - Fix race in Default ALC initialization.

### PR DESCRIPTION
**Pull request by gkhanna79**
Fixes https://github.com/dotnet/coreclr/issues/2881

@jkotas PTAL

### CODE REVIEW

**Comment by gkhanna79** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
LGTM

**Comment by gkhanna79** [comment]
@jashook @briansull Is the Windows Release build JIT test failure known issue?

http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/release_windows_nt_prtest/3092/testReport/

**Comment by gkhanna79** [comment]
@dotnet-bot test OSX x64 Release Build and Test

**Comment by gkhanna79** [comment]
@dotnet-bot test OSX x64 Release Build and Test please

**Comment by gkhanna79** [comment]
@dotnet-bot OSX x64 Release Build and Test

**Comment by mmitche** [comment]
@gkhanna79 You had it right the first and second time, it just may take a minute or so to launch.

**Comment by gkhanna79** [comment]
@mmitche do you know why the CI is so slow? NT, OSX and Ubuntu Release requests are running since morning :(

**Comment by stephentoub** [comment]
> NT, OSX and Ubuntu Release requests are running since morning :(

I believe some of the legs actually finished / timed out / failed, but there were failures (maybe due to the GitHub outage) in updating the summaries.  For example, you can see the OSX job actually failed while trying to clone the repo: http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/release_osx_flow_prtest/1103/.

File: src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 1]
```diff
@@ -396,6 +398,9 @@ public void StartProfileOptimization(string profile)
         // specified by the host. By having the field as a static, we are
         // making it an AppDomain-wide field.
         private static volatile AssemblyLoadContext m_DefaultAssemblyLoadContext;
+
+        // Synchronization primitive for controlling initialization of Default load context
+        private static object m_initLock = new Object();
```
Comment (stephentoub) [inline src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: Nit: ideally this would be readonly

Comment (stephentoub) [inline src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: (It's also a bit strange to me that ```m_``` is used as a prefix for statics, but I see that's already done elsewhere in the file.)

**Comment by stephentoub** [comment]
Test OSX x64 Release Build and Test please
Test Ubuntu x64 Release Build and Test please
Test Windows_NT x64 Release Build and Test please

**Comment by stephentoub** [comment]
LGTM

File: src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs (hunk 1 above)
Comment (gkhanna79) [inline src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: Thanks Stephen - I will incorporate your inputs - they seem reasonable :)

**Comment by mmitche** [comment]
@gkhanna79 For the Ubuntu and Windows legs, did you check status?  It's possible that we hit API limits again (I am talking with GitHub) and the status couldn't be updated on Github.  For OSX it's a lack of machines, which is being remedied.

We are also switching to an elastic pool of machines for Jenkins in general.

**Pull request merged**